MMM-SingleStock.js

```diff
@@ -152,5 +153,196 @@ Module.register('MMM-SingleStock', {
+        // eslint-disable-next-line prefer-destructuring
+        response = parsed[0];
+        break;
+      }
+      default:
+        break;
+    }
+
+    return response;
+  },
+
+  _setPrice(response) {
+    let value = '';
+    switch (this.config.api) {
+      case 'iexcloud':
+        value = response.latestPrice;
+        break;
+      case 'tiingo':
+        value = response.last;
+        break;
+      default:
+        break;
+    }
+
+    return value;
+  },
+
+  _setNumericChange(response) {
+    let value = '';
+    switch (this.config.api) {
+      case 'iexcloud':
+        value = response.change > 0 ? `+${response.change}` : `${response.change}`;
+        break;
+      case 'tiingo':
+        value = (response.prevClose - response.last).toFixed(2);
+        break;
+      default:
+        break;
+    }
+
+    return value;
+  },
+
+  _setPercentChange(response) {
+    let value = '';
+    switch (this.config.api) {
+      case 'iexcloud':
+        value = (response.changePercent * 100).toFixed(2);
+        break;
+      case 'tiingo':
+        value = (
+          ((response.prevClose - response.last) / response.prevClose)
+          * 100
+        ).toFixed(2);
+        break;
+      default:
+        break;
+    }
+
+    return value;
+  },
+
+  _SetUrl(api) {
+    switch (api) {
+      case 'iexcloud':
+        this.url = `https://cloud.iexapis.com/v1/stock/${this.config.stockSymbol}/quote?token=${this.config.apiToken}`;
+        break;
+      case 'tiingo':
+        if (this.url === '') {
+          this.url = this._getTiingoUrl();
+        }
+        break;
+      default:
+        Log.error(this.name, `Invalid api config: ${api}`);
+        break;
+    }
+  },
+
+  _getTiingoUrl() {
+    const expectedResponseHeaders = [
+      'server',
+      'date',
+      'content-type',
+      'content-length',
+      'vary',
+      'x-frame-options'
+    ];
+    const requestHeaders = [
+      { name: 'Content-Type', value: 'application/json' },
+      { name: 'Authorization', value: `Token ${this.config.apiToken}` }
+    ];
+
+    const url = this._getCorsUrl(
+      `https://api.tiingo.com/iex/?tickers=${this.config.stockSymbol}`,
+      requestHeaders,
+      expectedResponseHeaders
+    );
+
+    return url;
+  },
+
+  /**
+   * Gets a URL that will be used when calling the CORS-method on the server.
+   *
+   * @param {string} url the url to fetch from
+   * @param {Array.<{name: string, value:string}>} requestHeaders the HTTP headers to send
+   * @param {Array.<string>} expectedResponseHeaders the expected HTTP headers to recieve
+   * @returns {string} to be used as URL when calling CORS-method on server.
+   */
+  _getCorsUrl(url, requestHeaders, expectedResponseHeaders) {
+    if (!url || url.length < 1) {
+      throw new Error(`Invalid URL: ${url}`);
+    } else {
+      // eslint-disable-next-line no-restricted-globals
+      let corsUrl = `${location.protocol}//${location.host}/cors?`;
```

Comment:
🙋‍♂️ I'm not familiar with this endpoint, is this a CORS proxy service of MagicMirror? Does it really need for Tiingo, it does not support browsers?

Reply:
I'm honestly primarily a backend dev so dealing with CORS was a bit new for me but it was required for Tiingo.  I pretty much pulled the code from the [default magic mirror utils code](https://github.com/MichMich/MagicMirror/blob/master/modules/default/utils.js). I couldn't figure out how to directly use it.